src/lib/coaler/core/Matcher.cpp

```diff
@@ -0,0 +1,84 @@
+//
+// Created by malte on 12/4/23.
+//
+
+#include "Matcher.hpp"
+
+#include <GraphMol/RWMol.h>
+#include <GraphMol/SmilesParse/SmartsWrite.h>
+
+#include "GraphMol/ChemTransforms/ChemTransforms.h"
+#include "GraphMol/DistGeomHelpers/Embedder.h"
+#include "GraphMol/FMCS/FMCS.h"
+
+namespace coaler::core {
+    struct CompareUnsignedInts {
+        bool operator()(unsigned int a, unsigned int b) const { return a > b; }
+    };
+
+    std::optional<RDKit::ROMOL_SPTR> Matcher::calculateCoreMcs(RDKit::MOL_SPTR_VECT mols) {
+        RDKit::MCSParameters mcsParams;
+        RDKit::MCSAtomCompareParameters atomCompParams;
+        atomCompParams.MatchChiralTag = true;
+        atomCompParams.MatchFormalCharge = true;
+        atomCompParams.MatchIsotope = true;
+        atomCompParams.MatchValences = true;
+        atomCompParams.RingMatchesRingOnly = true;
+        atomCompParams.CompleteRingsOnly = true;
+        mcsParams.AtomCompareParameters = atomCompParams;
+
+        RDKit::MCSBondCompareParameters bondCompParams;
+        bondCompParams.MatchStereo = true;
+        bondCompParams.RingMatchesRingOnly = true;
+        bondCompParams.CompleteRingsOnly = true;
+        bondCompParams.MatchFusedRingsStrict = true;
+        mcsParams.BondCompareParameters = bondCompParams;
+
+        RDKit::MCSResult const mcs = RDKit::findMCS(mols, &mcsParams);
+        if (mcs.QueryMol == nullptr) {
+            return std::nullopt;
+        }
+
+        RDKit::RWMol first = *mols.at(0);
+
+        auto structMatches = RDKit::SubstructMatch(first, *mcs.QueryMol);
```

Comment:
Ich weiß ja du magst auto aber der return ist doch bestimmt wohldefiniert :)

Reply:
auto = "lieber compiler, bitte inserte für mich den return type, damit ich das nicht später händisch anpassen muss"

also der typ hier ist auch wohldefiniert und nicht irgendwie dynamisch oder so

Reply:
Ja genau warum nutzt du dann auto? :D

Reply:
Chris findet auto auch sinnvoll bei längeren Typ Bezeichnungen. Also hat er beim C++ Praktikum auch gesagt.

Reply:
Gründe warum ich gern auto benutze:
1. Der Typ ist weiter definiert und der Compiler macht die arbeit rauszfinden was da returned wird (i.e Ich verliere keine Informationen, weil ich immer in der IDE einfach sehen kann, welchen typ die variable hat, wenn ich das brauche)
2. Ich muss nicht für jeden `map<A, B> ...` typ (die sind ja echt common) irgendwo einen Typ alias definieren und den immer importen
3. Das ganze funktioniert auch mit lambdas und langen typen
4. Am wichtigsten: Wenn du mal irgendwo einen return-Typ ändern musst und nicht auto benutzt, musst du an allen stellen wo du die funktion callst den typ ändern. Das ist echt viel arbeit, bei einer größeren codebase.

Reply:
Ich finde es immer sinnvoll bei template typen oder sehr langen namen. Damit es die lesbarkeit nicht stört.